src/main/scala/vmu.scala

```diff
@@ -133,17 +133,17 @@ class IBox(implicit p: Parameters) extends VMUModule()(p) {
       _agent.io.id := io.id
       _agent.io.cfg <> io.cfg
       io.agu <> _agent.io.agu
-      _agent
-  } else Module(new IBoxSL)
+      _agent.io
```

Comment:
Is this necessary or just a style change that you prefer?

Reply:
Since chipsalliance/chisel3#1745 gets merged, `Module` doesn’t have `val io` any more, I think this modification is the minimal change version.
Here are another version: 
Define a trait `Has IBoxIO {val io = new IBoxIO}`, extends this trait for both `IBoxSL` and `IBoxML`, then we can use `io` field again. I personally think this will introduce too much changes to code base, so I directly pull `.io` out to avoid do this modification.

Reply:
You're [unintentionally] taking advantage of structural typing in Scala here. The type being inferred for `val agent` is `VMUModule { val io : Bundle { val id: UInt, val op: Decoupled[VMUOp], ... } }`, if you tried to give an explicit type you would either have to type all of this out or lose all `io` information. The best solution is what @sequencer is suggesting (but not PRing) which is to define a common `trait` that defines the common IO and then mix it in to each module class. The PR-ed change is just unwrapping the structural typing of `val agent` by one level to be `Bundle { val id: UInt, val op: Decoupled[VMUOp], ... }`

Reply:
Ok thanks for the explanation. I don't think its worth the refactoring now.